src/main/java/it/reply/orchestrator/service/deployment/providers/ChronosServiceImpl.java

```diff
@@ -247,7 +249,16 @@ public boolean isDeployed(DeploymentMessage deploymentMessage) {
 
   @Override
   public void cleanFailedDeploy(DeploymentMessage deploymentMessage) {
-    // DO NOTHING
+    doUndeploy(deploymentMessage);
```

Comment:
This do not respect the `isLastProvider` and `isKeepLastAttempt` logic

Reply:
the cycle of tests on the various providers is however carried out. keeplastattempt only makes sense for IM deployments

Reply:
>keeplastattempt only makes sense for IM deployments

why does it not?

Reply:
keeplastattempt serves when the deployment of large infrastructures is required in order to preserve a possible partial deployment (eg 150 VM out of 200), in other cases it is not needed

Reply:
it is also needed if you want to debug the partial deployment. But at the end of the day I think it is up to the user to decide whether to keep the last attempt or not. Plus keeping the mechanism will allow to maintain consistency between all the provider types

Reply:
moreover, the "keeplastattempt" logic is absolutely not implemented in those providers

Reply:
@maricaantonacci what do you think about it?

Reply:
> moreover, the "keeplastattempt" logic is absolutely not implemented in those providers

Reply:
I would say yes ... obviously if it was not needed it was not implemented (you should know it better than me since you probably did not)

Reply:
let me get this straight: do you need this because otherwise the workflow will continue on polling even if the timeout occurred?

Reply:
No, the process times out for various reasons but very likely because there are currently no resources available to complete the deployment; however, leaving the process running on the provider, it could be completed and therefore there would be a complete and untracked deployment (since the orchestrator is doing another one on the next provider). For this reason it is in any case necessary to abort the process on the provider and free up any resources already committed.

Reply:
please, can restart jenkins?

Reply:
> please, can restart jenkins?
that's not our jenkins but the EMT one. I do not have the credentials either

Reply:
>>>because otherwise the workflow will continue on polling even if the timeout occurred
handler throws BusinessWorkflowException...

Reply:
> No, the process times out for various reasons but very likely because there are currently no resources available to complete the deployment; however, leaving the process running on the provider, it could be completed and therefore there would be a complete and untracked deployment (since the orchestrator is doing another one on the next provider). For this reason it is in any case necessary to abort the process on the provider and free up any resources already committed.

Well, that's because the updated workflow is wrong. You're using an implicit parallel gateway, so the ProviderTimeout branch runs in "parallel" of the polling one. This is how it should be
![Screenshot (3)](https://user-images.githubusercontent.com/930764/69353985-b4af7900-0c7f-11ea-8217-36cdc4cc6a6e.png)

Reply:
very nice, commit the update...

Reply:
or even better, the inner subprocess could embrace both the "Deploy" and the "Poll for Deployment" tasks

Reply:
I believe that the deploy must remain external since the true goal of the timeout is not to check the time required for the deployment but that the system has actually started (think of a task that runs on marathon ...)